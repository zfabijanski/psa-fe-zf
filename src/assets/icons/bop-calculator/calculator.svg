<svg xmlns="http://www.w3.org/2000/svg" width="30" height="30" viewBox="0 0 30 30">
    <g fill="none" fill-rule="nonzero" transform="translate(8 6)">
        <path stroke="#222" stroke-width="2" d="M6.563 18h-5.25C.588 18 0 17.486 0 16.851V1.15C0 .514.588 0 1.313 0h11.374C13.412 0 14 .514 14 1.149V16a2 2 0 0 1-2 2H6.562z"/>
        <path stroke="#222" stroke-width="2" d="M1 1h12v5H1z"/>
        <rect width="2" height="2" x="3" y="8" fill="#222" rx="1"/>
        <rect width="2" height="2" x="3" y="11" fill="#222" rx="1"/>
        <rect width="2" height="2" x="6" y="8" fill="#222" rx="1"/>
        <rect width="2" height="2" x="6" y="11" fill="#222" rx="1"/>
        <rect width="2" height="2" x="9" y="8" fill="#222" rx="1"/>
        <rect width="2" height="2" x="9" y="11" fill="#222" rx="1"/>
        <rect width="2" height="2" x="3" y="14" fill="#222" rx="1"/>
        <rect width="2" height="2" x="6" y="14" fill="#222" rx="1"/>
        <rect width="2" height="2" x="9" y="14" fill="#222" rx="1"/>
    </g>
</svg>
