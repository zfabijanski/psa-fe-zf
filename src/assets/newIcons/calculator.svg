<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M14.2715 1.66699H4.89648C4.03354 1.66699 3.33398 2.41318 3.33398 3.33366V15.8337C3.33398 16.7541 4.03354 17.5003 4.89648 17.5003H14.2715C15.1344 17.5003 15.834 16.7541 15.834 15.8337V3.33366C15.834 2.41318 15.1344 1.66699 14.2715 1.66699Z" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
<rect x="5.08398" y="3.41699" width="9" height="4.83333" fill="#2F2F2F" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
<line x1="13.416" y1="9.91699" x2="12.416" y2="9.91699" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="13.416" y1="12.417" x2="12.416" y2="12.417" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="13.416" y1="14.083" x2="12.416" y2="14.083" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="13.416" y1="15.75" x2="12.416" y2="15.75" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="10.084" y1="9.91699" x2="9.08398" y2="9.91699" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="6.75" y1="9.91699" x2="5.75" y2="9.91699" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="10.084" y1="12.417" x2="9.08398" y2="12.417" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="6.75" y1="12.417" x2="5.75" y2="12.417" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="10.084" y1="14.083" x2="9.08398" y2="14.083" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="10.084" y1="15.75" x2="9.08398" y2="15.75" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="6.75" y1="14.083" x2="5.75" y2="14.083" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
<line x1="6.75" y1="15.75" x2="5.75" y2="15.75" stroke="#2F2F2F" stroke-width="1.5" stroke-linecap="round"/>
</svg>
