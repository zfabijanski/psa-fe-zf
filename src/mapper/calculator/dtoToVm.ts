import { get, keyBy, toString } from "lodash";
import moment from "moment";
import {
  ICalculationVM,
  ICover as ICoverVm,
  ICoversConfigMap,
  IFundContribution,
  IInsured as IInsuredVm,
  IProductCoversMap,
} from "../../features/Calculator/types";
import {
  Frequency,
  IFundRiskProfileConfig,
  IProductFund,
  IProductsConfigMap,
  ProductCategory,
  Role,
} from "../../models/common";
import {
  Direction,
  ICalculationDetailsDto,
  ICover as ICoverDto,
  IInsured as IInsuredDto,
  YesNo,
} from "models/calculator";
import { dateFormat } from "../../utils/dateCalculations";
import { roundToDecimals } from "utils/transformers";

const mapDtoToVm = (
  calculationDetailsDto: ICalculationDetailsDto,
  coversConfigMap: ICoversConfigMap,
  productCoversMap?: IProductCoversMap,
  productsConfigMap?: IProductsConfigMap,
  fundRiskProfiles?: IFundRiskProfileConfig[],
  mainCoverDirectionOfCalculation?: Direction
): ICalculationVM => {
  const isChildProduct = calculationDetailsDto.product_categories.includes(
    ProductCategory.Child
  );
  const productCovers =
    productCoversMap && calculationDetailsDto.product_guid
      ? productCoversMap[calculationDetailsDto.product_guid]
      : undefined;
  const productFunds =
    (productsConfigMap &&
      calculationDetailsDto.product_guid &&
      productsConfigMap[calculationDetailsDto.product_guid].product_funds) ||
    [];
  const mainCoverDto = calculationDetailsDto.covers_list.find(
    (cover) => cover.code === (productCovers ? productCovers.mainCover : "")
  );
  const frequencyCode =
    (mainCoverDto && mainCoverDto.frequency_code) ||
    (productsConfigMap &&
      productsConfigMap[calculationDetailsDto.product_guid].product_freqs[0]);

  const mainCoverConfig = productsConfigMap
    ? productsConfigMap[calculationDetailsDto.product_guid].product_covers.find(
        (cover) =>
          cover.cover_code === (productCovers ? productCovers.mainCover : "")
      )
    : undefined;

  const mainCover = mapCoverDtoToVm(
    productCovers ? productCovers.mainCover : "",
    false,
    mainCoverDto,
    false,
    mainCoverDirectionOfCalculation
  );
  const additionalCovers = mapCoversDtoToVm(
    calculationDetailsDto.covers_list,
    productCovers ? productCovers.additionalCovers : [],
    coversConfigMap,
    false
  );
  const wopCovers = mapCoversDtoToVm(
    calculationDetailsDto.covers_list,
    productCovers ? productCovers.wopCovers : [],
    coversConfigMap,
    true
  );

  const mainInsured = {
    ...mapInsuredDtoToVm(
      calculationDetailsDto.insureds_list.find(
        (insured) => insured.role === Role.MAIN || insured.role === Role.MAIN_PH
      )
    ),
    covers: isChildProduct
      ? mapCoversDtoToVm(
          calculationDetailsDto.covers_list,
          productCovers ? productCovers.childCovers : [],
          coversConfigMap,
          false,
          1
        )
      : [],
  };
  const policyHolder = mapInsuredDtoToVm(
    calculationDetailsDto.insureds_list.find(
      (insured) => insured.role === Role.PH || insured.role === Role.ALA_PH
    )
  );
  const additionalLifeAssureds: IInsuredVm[] =
    calculationDetailsDto.insureds_list
      .filter((insured) => insured.role === Role.ALA)
      .map((insured) => ({
        ...mapInsuredDtoToVm(insured),
        covers: mapCoversDtoToVm(
          calculationDetailsDto.covers_list,
          productCovers ? productCovers.childCovers : [],
          coversConfigMap,
          false,
          insured.insured_idx
        ),
      }));

  const fundContributions = mapFundsToVm(
    calculationDetailsDto,
    productFunds,
    fundRiskProfiles
  );

  return {
    calculationId: toValueOrDefault(calculationDetailsDto.calculation_id),
    productGuid: calculationDetailsDto.product_guid,
    productCategories: calculationDetailsDto.product_categories,
    premiumsTotal: toValueOrDefault(calculationDetailsDto.premiums_total),
    guaranteedDeathBenefit: toValueOrDefault(
      calculationDetailsDto.guaranteed_death_benefit
    ),
    guaranteedMaturityBenefit: toValueOrDefault(
      calculationDetailsDto.guaranteed_maturity_benefit
    ),
    projectedMaturityValueEgrM: toValueOrDefault(
      calculationDetailsDto.projected_maturity_value_egr_m
    ),
    policyHolderIsMainInsured: toYesNo(
      calculationDetailsDto.insureds_list.some(
        (insured) => insured.role === Role.MAIN_PH
      )
    ),
    indexation: toValueOrDefault(mainCoverDto ? mainCoverDto.indexation : null),
    ropOption: toValueOrDefault(mainCoverDto ? mainCoverDto.rop_option : null),
    ropAvailable: toValueOrDefault(
      mainCoverConfig ? mainCoverConfig.rop_available : null
    ),
    investmentProfile: toValueOrDefault(
      mainCoverDto ? mainCoverDto.investment_profile : null
    ),
    frequencyCode: frequencyCode as Frequency,
    mainCover,
    additionalCovers,
    wopCovers,
    mainInsured,
    policyHolder,
    additionalLifeAssureds,
    fundContributions,
  };
};

const mapCoversDtoToVm = (
  covers: ICoverDto[],
  coversCode: string[],
  coversConfigMap: ICoversConfigMap,
  isWOP: boolean,
  insuredIndex?: number
): ICoverVm[] => {
  return coversCode.map((coverCode) =>
    mapCoverDtoToVm(
      coverCode,
      isWOP,
      covers.find(
        (cover) =>
          cover.code === coverCode &&
          (!insuredIndex || cover.insured_idx === insuredIndex)
      ),
      coversConfigMap[coverCode].isHospital
    )
  );
};

const mapInsuredDtoToVm = (insuredDto?: IInsuredDto): IInsuredVm => {
  if (!!insuredDto) {
    return {
      insuredUuid: toValueOrDefault(insuredDto.insured_uuid),
      dateOfBirth: toDateStringOrDefault(insuredDto.date_of_birth),
      covers: [],
    };
  } else {
    return {
      covers: [],
    };
  }
};

export const mapCoverDtoToVm = (
  coverCode: string,
  isWOP: boolean,
  coverDto?: ICoverDto,
  isHospital?: boolean,
  directionOfCalculation?: Direction
): ICoverVm => {
  if (!!coverDto) {
    return {
      checked: YesNo.Yes,
      code: coverCode,
      payments: toValueOrDefault(coverDto.payments),
      directionOfCalculation:
        directionOfCalculation ||
        toValueOrDefault(
          coverDto.direction_of_calculation,
          isWOP ? Direction.SumAssuredToPremium : undefined
        ),
      sumAssured: !isWOP ? toValueOrDefault(coverDto.sum_assured) : undefined,
      premiumPerFreq: toValueOrDefault(coverDto.premium_per_freq),
      duration: toValueOrDefault(coverDto.duration),
      idxOption: toValueOrDefault(
        coverDto.idx_option,
        isWOP ? YesNo.No : undefined
      ),
      addSumAssured: toValueOrDefault(
        isHospital
          ? coverDto.sum_assured && roundToDecimals(coverDto.sum_assured / 100)
          : coverDto.add_sum_assured
      ),
      expMatBen: toValueOrDefault(coverDto.exp_mat_ben),
      sumOfPremiums:
        coverDto.payments !== null && coverDto.premium_per_freq !== null
          ? coverDto.payments * coverDto.premium_per_freq
          : undefined,
    };
  } else {
    return {
      checked: YesNo.No,
      code: coverCode,
      directionOfCalculation: isWOP ? Direction.SumAssuredToPremium : undefined,
      idxOption: isWOP ? YesNo.No : undefined,
    };
  }
};

export const mapFundsToVm = (
  calculationDetailsDto: ICalculationDetailsDto,
  productFunds: IProductFund[],
  fundRiskProfiles: IFundRiskProfileConfig[] = []
): IFundContribution[] => {
  const calculatorFunds = keyBy(calculationDetailsDto.funds_list, "fund_code");

  return fundRiskProfiles.map((fundRiskProfile) => ({
    code: fundRiskProfile.code,
    nameEN: fundRiskProfile.name_en,
    namePL: fundRiskProfile.name_pl,
    notes: fundRiskProfile.notes,
    riskLevel: fundRiskProfile.risk_level,
    statusCode: fundRiskProfile.status_code,
    funds: productFunds
      .filter(
        (productFund) => productFund.fund_risk_profile === fundRiskProfile.code
      )
      .map((fund) => {
        const { guarantee_code, guarantee_name, guarantee_percentages } =
          fund.guarantees[0] ?? {};
        const guaranteePercentageAvailable =
          guarantee_percentages?.map(({ guarantee_percent }) =>
            String(guarantee_percent)
          ) ?? [];

        return {
          coverIdx: 1,
          fundCode: fund.fund_code,
          fundName: fund.fund_name,
          fundRiskProfile: fund.fund_risk_profile,
          fundOrder: fund.fund_order,
          allocationValue: toString(
            get(calculatorFunds, [fund.fund_code, "allocation_value"])
          ),
          allocationAmount: toString(
            get(calculatorFunds, [fund.fund_code, "allocation_amount"])
          ),
          guarantee: {
            guaranteeCode: guarantee_code,
            guaranteeName: guarantee_name,
            guaranteePercentageAvailable,
            guaranteeChecked: !!get(calculatorFunds, [
              fund.fund_code,
              "fund_gtees_list",
              "0",
            ])?.allocation_percent,
            guaranteeValue: toString(
              get(calculatorFunds, [fund.fund_code, "fund_gtees_list", "0"])
                ?.allocation_percent
            ),
          },
        };
      }),
  }));
};

export const toValueOrDefault = <T extends {}>(
  value: T | null,
  defaultValue?: T
): T | undefined => (value !== null ? value : defaultValue);

const toDateStringOrDefault = (
  value: string | null,
  defaultValue?: string
): string | undefined =>
  value !== null && moment(value).isValid()
    ? moment(value).format(dateFormat)
    : defaultValue;

const toYesNo = (value?: boolean) => (value ? YesNo.Yes : YesNo.No);

export default mapDtoToVm;
