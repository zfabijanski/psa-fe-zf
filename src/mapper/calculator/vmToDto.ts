import moment from "moment";
import {
  ICalculationVM,
  ICover as ICoverVm,
  IFundContribution,
  IInsured as IInsuredVm,
} from "../../features/Calculator/types";
import {
  Frequency,
  InvestmentProfile,
  ProductCategory,
  Role,
} from "../../models/common";
import {
  Direction,
  ICalculationDto,
  ICover as ICoverDto,
  IFund as IFundDto,
  IInsured as IInsuredDto,
  YesNo,
} from "models/calculator";
import { dateFormat } from "../../utils/dateCalculations";
import { Indexer } from "../../utils/indexer";

const mapVmToDto = (calculationVM: ICalculationVM): ICalculationDto => {
  const coversIndexer = new Indexer(1);
  const insuredsIndexer = new Indexer(1);

  const mainInsured = mapInsuredVmToDto(
    calculationVM.mainInsured,
    calculationVM.policyHolderIsMainInsured === YesNo.Yes
      ? Role.MAIN_PH
      : Role.MAIN,
    insuredsIndexer.getNext()
  );
  const mainCover = mapCoverVmToDto(
    calculationVM.mainCover,
    coversIndexer.getNext(),
    insuredsIndexer.lastIndex,
    calculationVM.frequencyCode,
    calculationVM.productCategories.includes(ProductCategory.Child),
    calculationVM.indexation,
    calculationVM.ropOption,
    calculationVM.investmentProfile
  );
  const additionalCovers = mapCoversVmToDto(
    calculationVM.additionalCovers,
    coversIndexer,
    insuredsIndexer.lastIndex,
    calculationVM.frequencyCode
  );
  const mainInsuredChildCovers = mapCoversVmToDto(
    calculationVM.mainInsured.covers,
    coversIndexer,
    insuredsIndexer.lastIndex,
    calculationVM.frequencyCode
  );
  const policyHolder =
    calculationVM.policyHolderIsMainInsured !== YesNo.Yes
      ? mapInsuredVmToDto(
          calculationVM.policyHolder,
          (calculationVM.wopCovers || []).some(
            (cover) => cover.checked === YesNo.Yes
          ) && calculationVM.productCategories.includes(ProductCategory.Child)
            ? Role.ALA_PH
            : Role.PH,
          insuredsIndexer.getNext()
        )
      : undefined;
  const wopCovers = mapCoversVmToDto(
    calculationVM.wopCovers,
    coversIndexer,
    insuredsIndexer.lastIndex,
    calculationVM.frequencyCode
  );
  const additionalLifeAssureds: IInsuredDto[] = [];
  const additionalLifeAssuredsChildCovers: ICoverDto[] = [];
  (calculationVM.additionalLifeAssureds || []).forEach((child) => {
    additionalLifeAssureds.push(
      mapInsuredVmToDto(child, Role.ALA, insuredsIndexer.getNext())
    );
    additionalLifeAssuredsChildCovers.push(
      ...mapCoversVmToDto(
        child.covers,
        coversIndexer,
        insuredsIndexer.lastIndex,
        calculationVM.frequencyCode
      )
    );
  });

  const covers: ICoverDto[] = [
    mainCover,
    ...additionalCovers,
    ...mainInsuredChildCovers,
    ...wopCovers,
    ...additionalLifeAssuredsChildCovers,
  ];
  const insureds: IInsuredDto[] = [
    mainInsured,
    ...(policyHolder ? [policyHolder] : []),
    ...additionalLifeAssureds,
  ];

  const funds: IFundDto[] = mapFundsVmToDto(calculationVM.fundContributions);

  return {
    uuid: "123", // TODO: Set uuid, version and source_system properly
    version: "456",
    source_system: "FE",
    product_guid: calculationVM.productGuid,
    covers_list: covers,
    insureds_list: insureds,
    funds_list: funds,
  };
};

const mapCoversVmToDto = (
  coversVm: ICoverVm[] | undefined,
  indexer: Indexer,
  insuredIndex: number,
  frequencyCode?: Frequency
): ICoverDto[] => {
  return (coversVm || [])
    .filter((cover) => cover.checked === YesNo.Yes)
    .map((cover) =>
      mapCoverVmToDto(cover, indexer.getNext(), insuredIndex, frequencyCode)
    );
};

const mapInsuredVmToDto = (
  insuredVm: IInsuredVm,
  role: Role,
  insuredIndex: number
): IInsuredDto => {
  return {
    insured_idx: insuredIndex,
    insured_uuid: toValueOrNull(insuredVm.insuredUuid),
    age_in_months: null,
    date_of_birth: toISOStringOrNull(insuredVm.dateOfBirth),
    role,
  };
};

const mapCoverVmToDto = (
  coverVm: ICoverVm,
  coverIndex: number,
  insuredIndex: number,
  frequencyCode?: Frequency,
  isChildMainCover?: boolean,
  indexation?: YesNo,
  ropOption?: YesNo,
  investmentProfile?: InvestmentProfile
): ICoverDto => {
  return {
    cover_idx: coverIndex,
    insured_idx: insuredIndex,
    code: coverVm.code,
    frequency_code: toValueOrNull(frequencyCode),
    payments: toValueOrNull(coverVm.payments),
    direction_of_calculation: toValueOrNull(coverVm.directionOfCalculation),
    investment_profile: toValueOrNull(investmentProfile),
    indexation: toValueOrNull(indexation),
    sum_assured: toValueOrNull(
      coverVm.sumAssured,
      coverVm.directionOfCalculation === Direction.SumAssuredToPremium
    ),
    premium_per_freq: toValueOrNull(
      coverVm.premiumPerFreq,
      coverVm.directionOfCalculation === Direction.PremiumToSumAssured
    ),
    duration: toValueOrNull(coverVm.duration),
    idx_option: toValueOrNull(coverVm.idxOption),
    rop_option: toValueOrNull(ropOption),
    add_sum_assured: toValueOrNull(
      coverVm.addSumAssured,
      isChildMainCover ||
        coverVm.directionOfCalculation === Direction.SumAssuredToPremium
    ),
    ending_age: null,
    exp_mat_ben: toValueOrNull(
      coverVm.expMatBen,
      coverVm.directionOfCalculation === Direction.MaturityBenefitToPremium
    ),
  };
};

export const mapFundsVmToDto = (
  fundContributions: Array<Partial<IFundContribution>>
): IFundDto[] => {
  const funds = fundContributions.map((x) => x.funds).flat();

  return funds.map((fund) => ({
    cover_idx: fund?.coverIdx || 1,
    allocation_value: Number(fund?.allocationValue) || 0,
    fund_code: toValueOrNull(fund?.fundCode),
    fund_risk_profile: toValueOrNull(fund?.fundRiskProfile),
    fund_gtees_list: fund?.guarantee?.guaranteeChecked
      ? [
          {
            allocation_percent: Number(fund?.guarantee.guaranteeValue),
            gtee_code: fund?.guarantee.guaranteeCode,
          },
        ]
      : null,
  }));
};

const toValueOrNull = <T extends {}>(
  value: T | undefined,
  condition: boolean = true
) => (condition ? value || null : null);

const toISOStringOrNull = (value?: string) =>
  value ? moment(value, dateFormat).utc(true).toISOString() : null;

export default mapVmToDto;
