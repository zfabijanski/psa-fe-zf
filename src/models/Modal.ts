import { ReactNode } from "react";
import { TIconType } from "../components/UI/Icon";
import { TranslationValue } from "utils/types";

export enum ButtonTypes {
  cancel = "cancel",
  confirm = "confirm",
}

export enum ModalTypes {
  default = "default",
  danger = "danger",
}

export interface IModalData {
  modalIllustration?: TIconType;
  modalTitleTrKey?: string;
  modalContentTrKey?:
    | string
    | (
        | string
        | string[]
        | { trKey: string; values?: Record<string, ReactNode> }
      )[];
  ModalComponent?: React.FC;
  modalContent?: ReactNode;
  modalHtml?: string;
  modalHtmlTrValues?: Record<string, TranslationValue>;
  modalButtons?: {
    [key in ButtonTypes]?: {
      textTrKey: string;
      onClick?: () => void;
    };
  };
  modalType?: ModalTypes;
  maskClosable?: boolean;
}
