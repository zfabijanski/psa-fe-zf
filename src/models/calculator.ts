import {
  Frequency,
  InvestmentProfile,
  ProductCategory,
  Role,
} from "models/common";

export type CalculationState = {
  status?: CalculationStatus;
  calculationFeatureChanged: boolean;
};

export interface ICalculationDetailsDto extends ICalculationDto {
  calculation_id: number;
  working_date: string;
  premiums_total: number;
  guaranteed_death_benefit: number | null;
  guaranteed_maturity_benefit: number | null;
  projected_maturity_value_egr_m: number | null;
  product_categories: ProductCategory[];
}

export interface ICalculateResponseDto {
  calculation: ICalculationDetailsDto;
  msg: IMessage[];
}

interface IMessage {
  msg_code: string;
  msg_type: MessageType;
  msg_mesage: string;
  msg_function: string;
  msg_key?: string;
}

export enum MessageType {
  Ok = "OK",
  Error = "ERROR",
}

export interface ICalculationDto {
  uuid: string | null;
  version: string | null;
  source_system: string | null;
  product_guid: number;
  insureds_list: IInsured[];
  covers_list: ICover[];
  funds_list: IFund[];
}

export interface IInsured {
  insured_idx: number;
  insured_uuid: string | null;
  age_in_months: number | null;
  date_of_birth: string | null;
  role: Role;
}

export interface ICover {
  cover_id?: number;
  cover_idx: number;
  insured_idx: number;
  code: string;
  frequency_code: Frequency | null;
  payments: number | null;
  direction_of_calculation: Direction | null;
  investment_profile: InvestmentProfile | null;
  indexation: YesNo | null;
  sum_assured: number | null;
  premium_per_freq: number | null;
  duration: number | null;
  idx_option: YesNo | null;
  rop_option: YesNo | null;
  add_sum_assured: number | null;
  ending_age: number | null;
  exp_mat_ben: number | null;
}

export interface IFundGuarantee {
  gtee_code: string;
  allocation_percent: number;
}

export interface IFund {
  cover_idx: number;
  fund_code: string | null;
  fund_name?: string | null;
  fund_risk_profile: string | null;
  allocation_value: number | null;
  fund_gtees_list?: IFundGuarantee[] | null;
}

export enum Direction {
  PremiumToSumAssured = "PREM_TO_SA",
  SumAssuredToPremium = "SA_TO_PREM",
  MaturityBenefitToPremium = "MATBEN_TO_PREM",
}

export enum YesNo {
  Yes = "Y",
  No = "N",
}

export enum CalculationStatus {
  NotCalculated = "NOT_CALCULATED",
  Invalid = "INVALID",
  Calculated = "CALCULATED",
  Illustrated = "ILUSTRATED",
}
