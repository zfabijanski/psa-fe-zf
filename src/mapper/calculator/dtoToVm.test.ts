import { Role } from "models/common";
import mapDtoToVm, { mapFundsToVm, mapCoverDtoToVm } from "./dtoToVm";
import {
  createCalculationDetailsDto,
  createCalculatorCoversConfigMap,
  createProductConfigMap,
  createFundRiskProfileConfig,
  createProductCoversMap,
  createInsured,
  createProductFund,
  createCover,
} from "testUtils";
import { Direction } from "models/calculator";

describe("Mapping DTO to VM", () => {
  it("maps DTO to VM", () => {
    expect(
      mapDtoToVm(
        createCalculationDetailsDto(),
        createCalculatorCoversConfigMap(),
        createProductCoversMap(),
        createProductConfigMap(),
        [createFundRiskProfileConfig()]
      )
    ).toMatchSnapshot();
  });

  it('maps fund risk profile to "low" when fund risk profile is not found', () => {
    expect(
      mapDtoToVm(
        createCalculationDetailsDto(),
        createCalculatorCoversConfigMap(),
        createProductCoversMap(),
        createProductConfigMap(),
        []
      )
    ).toMatchSnapshot();
  });

  it("formats date of birth for insured person", () => {
    expect(
      mapDtoToVm(
        createCalculationDetailsDto({
          insureds_list: [
            createInsured({ date_of_birth: "2020-01-01", role: Role.MAIN }),
          ],
        }),
        createCalculatorCoversConfigMap(),
        createProductCoversMap(),
        createProductConfigMap(),
        [createFundRiskProfileConfig()]
      )
    ).toHaveProperty("mainInsured", {
      covers: [],
      dateOfBirth: "2020-01-01",
      insuredUuid: "insured_uuid",
    });
  });

  it("formats date of birth for policy holder", () => {
    expect(
      mapDtoToVm(
        createCalculationDetailsDto({
          insureds_list: [
            createInsured({ date_of_birth: "2020-01-01", role: Role.PH }),
          ],
        }),
        createCalculatorCoversConfigMap(),
        createProductCoversMap(),
        createProductConfigMap(),
        [createFundRiskProfileConfig()]
      )
    ).toHaveProperty("policyHolder", {
      covers: [],
      dateOfBirth: "2020-01-01",
      insuredUuid: "insured_uuid",
    });
  });

  it("formats date of birth for additional life assured", () => {
    expect(
      mapDtoToVm(
        createCalculationDetailsDto({
          insureds_list: [
            createInsured({ date_of_birth: "2020-01-01", role: Role.ALA }),
          ],
        }),
        createCalculatorCoversConfigMap(),
        createProductCoversMap(),
        createProductConfigMap(),
        [createFundRiskProfileConfig()]
      )
    ).toHaveProperty("additionalLifeAssureds", [
      {
        covers: [],
        dateOfBirth: "2020-01-01",
        insuredUuid: "insured_uuid",
      },
    ]);
  });
});

describe("Mapping funds to VM", () => {
  it("maps funds to VM", () => {
    expect(
      mapFundsToVm(
        createCalculationDetailsDto(),
        [createProductFund()],
        [createFundRiskProfileConfig()]
      )
    ).toMatchInlineSnapshot(`
      Array [
        Object {
          "code": "code",
          "funds": Array [
            Object {
              "allocationAmount": "",
              "allocationValue": "",
              "coverIdx": 1,
              "fundCode": "fund_code",
              "fundName": "fund_name",
              "fundOrder": 1,
              "fundRiskProfile": "code",
              "guarantee": Object {
                "guaranteeChecked": false,
                "guaranteeCode": "guarantee_code",
                "guaranteeName": "guarantee_name",
                "guaranteePercentageAvailable": Array [
                  "10",
                ],
                "guaranteeValue": "",
              },
            },
          ],
          "nameEN": "name_en",
          "namePL": "name_pl",
          "notes": "notes",
          "riskLevel": 1,
          "statusCode": "status_code",
        },
      ]
    `);
  });
});

describe("Mapping covers to VM", () => {
  it('maps cover "amount" to VM', () => {
    expect(
      mapCoverDtoToVm(
        "cover_code",
        true,
        createCover(),
        false,
        Direction.SumAssuredToPremium
      )
    ).toMatchInlineSnapshot(`
      Object {
        "addSumAssured": 1,
        "checked": "Y",
        "code": "cover_code",
        "directionOfCalculation": "SA_TO_PREM",
        "duration": 1,
        "expMatBen": 1,
        "idxOption": "Y",
        "payments": 1,
        "premiumPerFreq": 1,
        "sumAssured": undefined,
        "sumOfPremiums": 1,
      }
    `);
  });
});
