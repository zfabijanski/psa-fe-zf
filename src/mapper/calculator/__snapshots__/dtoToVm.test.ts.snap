// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mapping DTO to VM maps DTO to VM 1`] = `
Object {
  "additionalCovers": Array [],
  "additionalLifeAssureds": Array [],
  "calculationId": 1,
  "frequencyCode": undefined,
  "fundContributions": Array [
    Object {
      "code": "code",
      "funds": Array [],
      "nameEN": "name_en",
      "namePL": "name_pl",
      "notes": "notes",
      "riskLevel": 1,
      "statusCode": "status_code",
    },
  ],
  "guaranteedDeathBenefit": 100,
  "guaranteedMaturityBenefit": 100,
  "indexation": undefined,
  "investmentProfile": undefined,
  "mainCover": Object {
    "checked": "N",
    "code": "",
    "directionOfCalculation": undefined,
    "idxOption": undefined,
  },
  "mainInsured": Object {
    "covers": Array [],
  },
  "policyHolder": Object {
    "covers": Array [],
  },
  "policyHolderIsMainInsured": "N",
  "premiumsTotal": 100,
  "productCategories": Array [],
  "productGuid": 1,
  "projectedMaturityValueEgrM": 100,
  "ropAvailable": undefined,
  "ropOption": undefined,
  "wopCovers": Array [],
}
`;

exports[`Mapping DTO to VM maps fund risk profile to "low" when fund risk profile is not found 1`] = `
Object {
  "additionalCovers": Array [],
  "additionalLifeAssureds": Array [],
  "calculationId": 1,
  "frequencyCode": undefined,
  "fundContributions": Array [],
  "guaranteedDeathBenefit": 100,
  "guaranteedMaturityBenefit": 100,
  "indexation": undefined,
  "investmentProfile": undefined,
  "mainCover": Object {
    "checked": "N",
    "code": "",
    "directionOfCalculation": undefined,
    "idxOption": undefined,
  },
  "mainInsured": Object {
    "covers": Array [],
  },
  "policyHolder": Object {
    "covers": Array [],
  },
  "policyHolderIsMainInsured": "N",
  "premiumsTotal": 100,
  "productCategories": Array [],
  "productGuid": 1,
  "projectedMaturityValueEgrM": 100,
  "ropAvailable": undefined,
  "ropOption": undefined,
  "wopCovers": Array [],
}
`;
